ra/ra.go

```diff
@@ -1563,24 +1569,54 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 		return nil, err
 	}
 
-	gotAuthzFor := make(map[string]bool, len(order.Names))
+	// Collect up the authorizations we found into a map keyed by the domains the
+	// authorizations correspond to
+	nameToExistingAuthz := make(map[string]*corepb.Authorization, len(orderIdentifiers))
 	for _, v := range existingAuthz.Authz {
-		gotAuthzFor[*v.Domain] = true
-		order.Authorizations = append(order.Authorizations, *v.Authz.Id)
+		nameToExistingAuthz[*v.Domain] = v.Authz
+	}
+
+	// For each of the names in the order, if there is an acceptable
+	// existing authz, append its ID to the order to reuse it. Otherwise track
+	// that there is a missing authz for a name
+	var missingAuthzs []core.AcmeIdentifier
```

Comment:
Since the only part of the identifier that is used is the name this can just be a `[]string`.

Reply:
If I make this change then I have to create a throw-away identifier out of the name again for `ra.createPendingAuthz` - it seems _slightly_ better to use the identifiers that are already constructed instead of pulling out the value and then making it an identifier again (or changing `ra.createPendingAuthz`).

Reply:
:+1:.